src/GafferOSL/ShadingEngine.cpp

```diff
@@ -191,6 +195,21 @@ class RenderState
 
 		bool userData( ustring name, TypeDesc type, void *value ) const
 		{
+			if( name == gIndex )
+			{
+				// if a 4 byte type has been requested then ensure we fit and cast to narrower type
+				// this way f32 reads of _index will succeed.
+				if( type.size() == sizeof( int ) && m_pointIndex <= ( (size_t) std::numeric_limits<int>::max() ) )
+				{
+					int v = ( int ) m_pointIndex;
+					return ShadingSystem::convert_value( value, type, &v, OIIO::TypeDesc( OIIO::TypeDesc::INT32 ) );
+				}
+				else
+				{
+					return ShadingSystem::convert_value( value, type, &m_pointIndex, OIIO::TypeDesc( OIIO::TypeDesc::UINT64 ) );
```

Comment:
Is there an OSL query where the UINT64 code path works? If this path is basically non-useful, I'm wondering if it might be better to always use the Int32 bit path, but issue a warning on overflow, rather than have a silent failure.